libbpf-tools/biolatency.bpf.c

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "biolatency.h"
+#include "bits.bpf.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile char targ_disk[DISK_NAME_LEN] = {};
+const volatile bool targ_per_disk = false;
+const volatile bool targ_per_flag = false;
+const volatile bool targ_queued = false;
+const volatile bool targ_ms = false;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} start SEC(".maps");
+
+static struct hist initial_hist;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct hist_key);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static __always_inline
+int trace_rq_start(struct request *rq)
+{
+	u64 ts = bpf_ktime_get_ns();
+	char disk[DISK_NAME_LEN];
+
+	BPF_CORE_READ_STR_INTO(&disk, rq, rq_disk, disk_name);
+	if (!str_filtered(disk, targ_disk, DISK_NAME_LEN))
+		return 0;
+
+	bpf_map_update_elem(&start, &rq, &ts, 0);
+	return 0;
+}
+
+SEC("tp_btf/block_rq_insert")
+int BPF_PROG(tp_btf__block_rq_insert, struct request_queue *q,
+	     struct request *rq)
+{
+	return trace_rq_start(rq);
+}
+
+SEC("tp_btf/block_rq_issue")
+int BPF_PROG(tp_btf__block_rq_issue, struct request_queue *q,
+	     struct request *rq)
+{
+	if (targ_queued && BPF_CORE_READ(q, elevator))
```

Comment:
it's a tp_btf program, so you don't have to use BPF_CORE_READ macro, just `q->elevator` will work and will be faster.

Reply:
Yes, @anakryiko I used `if (targ_queued && q->elevator)` first, but I found I always get nothing output (with none I/O scheduler: `echo none > /sys/block/sdc/queue/scheduler`, which set `q->elevator = NULL;`), but with `BPF_CORE_READ` is ok.

The kernel I tested is on `5.7.0-rc7+` and `5.8.0-rc1+`, both have this problem. I also found if the struct's member's type is pointer,  `if(struct->pointer_member)` will not work. So is this a bug or am I missed sth?

The workload source is 

```go
package main

import (
        "flag"
        "log"
        "os"
        "time"
)

func main() {
        name := flag.String("file", "", "filename to use")
        size := flag.Int("size", 1024*10, "chunk size")
        sync := flag.Int("sync", 1, "sync file after that many writes, <= 0 to disable")
        sleep := flag.Duration("sleep", time.Second, "time to sleep between writes")
        count := flag.Int("count", 0, "number of iterations, <=0 for infinity")

        flag.Parse()

        if *name == "" {
                flag.PrintDefaults()
        }

        file, err := os.OpenFile(*name, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
        if err != nil {
                log.Fatalf("Error opening file %q: %s", *name, err)
        }

        buf := make([]byte, *size)

        i := 0
        for {
                i++

               started := time.Now()

                verb := "Written"

                if _, err := file.Write(buf); err != nil {
                        log.Fatalf("Error writing: %s", err)
                }     
                if *sync > 0 && i%*sync == 0 {
                        if err := file.Sync(); err != nil {
                                log.Fatalf("Error syncing: %s", err)
                        }

                        verb = "Synced"
                }

                elapsed := time.Since(started)

                log.Printf("%s in %d μs | %v", verb, int(elapsed.Seconds() * 1e6), elapsed)

                if *count > 0 && i >= *count {
                        break
                }

                time.Sleep(*sleep)
        }
}
```

Then run:

`./biolatency   -d sdc  -QDF`
`go run ./paranoid.go -file=/dev/sdc -size=64 -count=10.   (write disk without fs)`


WIth 

`if (targ_queued && BPF_CORE_READ(q, elevator))`

I can always get

```
disk = sdc	flags = Sync-Write
     usecs               : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 8        |                                        |
        32 -> 63         : 2        |****************************************|
```

but with

`if(struct->pointer_member)`

Output nothing.

Reply:
It does sound like a bug. Can you please create a smallest repro and submit it to bpf@vger.kernel.org? Please also attach a BPF assembly dump ( with `llvm-objdump -d`). Thanks a lot for your help!

Reply:
Sure, you're welcome.

Reply:
@anakryiko Has reported, the subject is `tp_btf: if (!struct->pointer_member) always actually false although pointer_member == NULL`